File: Data Structures & Algorithms/combination-target-sum/submission-2.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        """
            Combination Sum

            [2, 5, 6, 9]
            target = 9

            for loop
        """

        result = []

        def dfs(idx: int, target: int, path: List[int]):
            if target < 0 or idx >= len(nums):
                return
            
            if target == 0:
                result.append(path.copy())
                return
            

            path.append(nums[idx])
            dfs(idx, target-nums[idx], path)
            path.pop()


            dfs(idx+1, target, path)


        def dfs1(idx: int, target: int, path: List[int]):
            if target < 0 or idx >= len(nums):
                return
            
            if target == 0:
                result.append(path.copy())
                return
            
            for i in range(idx,len(nums)):
                if target - nums[i] < 0:
                    continue
                
                path.append(nums[i])
                dfs1(i, target - nums[i], path)
                path.pop()
            

        #dfs(0,target,[])
        dfs1(0,target,[])
        return result

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-2.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum = accumulate(nums.begin(),nums.end(),0);
        if(sum % 2 == 1){
            return false;
        }

        int target = sum/2;
        int n = nums.size();
        vector<vector<int>>dp(n+1,vector<int>(target+1,0));
        dp[0][0] = 1;

        for(int i=1; i<=n; i++){
            for(int j=1; j<=target; j++){
                if(j >= nums[i-1]){
                    dp[i][j] = dp[i-1][j] || dp[i-1][j - nums[i-1]]; 
                }
                else{
                    dp[i][j] = dp[i-1][j];
                }
            }
        }
        return dp[n][target];
    }
};
